stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -285,6 +285,9 @@ mul!(out::AbstractMatrix, A::Diagonal, in::StridedMatrix) = out .= A.diag .* in
 mul!(out::AbstractMatrix, A::Adjoint{<:Any,<:Diagonal}, in::StridedMatrix) = out .= adjoint.(A.parent.diag) .* in
 mul!(out::AbstractMatrix, A::Transpose{<:Any,<:Diagonal}, in::StridedMatrix) = out .= transpose.(A.parent.diag) .* in
 
+# add methods for mul! when right multiplication by diagonal (no need for an AbstractVector case)
+mul!(out::AbstractMatrix, in::StridedMatrix, A::Diagonal) = out .= transpose(A.diag) .* in
```

Comment:
I'm not sure this is correct for noncommutative element types (like Quaternions, or matrices). I think it's better to keep the order of multiplication, so
```
mul!(out::AbstractMatrix, in::StridedMatrix, A::Diagonal) = out .= in .* transpose(A.diag)
```
consistently with line 175 above. Also, why not add the adjoint and transpose cases consistently with the ones a few lines above?
```
mul!(out::AbstractMatrix, in::StridedMatrix, A::Adjoint{<:Any,<:Diagonal}) = out .= in .* adjoint(A.parent.diag)
mul!(out::AbstractMatrix, in::StridedMatrix, A::Transpose{<:Any,<:Diagonal}) = out .= in .* transpose(A.parent.diag)
```
Note that `adjoint` and `transpose` act recursively, so they make the `diag` vector horizontal and adjoint/transpose the elements.

Reply:
@raphbacher You're right, `transpose` and `adjoint` on `Diagonal`s don't give `Adjoint{<:Any,<:Diagonal}` according to lines 437-440. Which is interesting, because there are a couple of methods defined for `Adjoint{<:Any,<:Diagonal}`. These, however, can only be constructed manually by calling the constructor `Adjoint(D)`. So, I assume that we should have them for completeness, even though they may be "never" called?